### PR TITLE
release: fetch release blockers

### PR DESCRIPTION
**Pull request by celiala**
This commit implements the Week -1 task of fetching the list of open release blockers for the specified releaseSeries,
and summarizing and sorting the blocker list by project name (if any).

#### Implementation Notes

This PR associates a release blocker with the most recently-added project, so that we report the correct number of open blockers.

If you're using roashdash to verify open blocker data, note that https://roachdash.crdb.dev/  groups things by project, which means that if an issue belongs to multiple projects, that issue will appear under both projects. (I.e. the old release process was sometimes double-counting and over reporting the number of open blockers, if an issue was assigned to multiple project).

### Manual Testing

Confirmed blockers for older production release branch (21.2) and upcoming 22.1 release (which as on March 9 is still `master`, since the 22.1 branch hasn't been cut yet).

```
# Test 1: generates alpha email
# --next-version v22.1.0-alpha.3
# --release-series 22.1
#
SMTP_PASSWORD=$MY_SMTP_PWD SMTP_USER=$MY_SMTP_USER GITHUB_TOKEN=$MY_GITHUB_TOKEN \
$(bazel info bazel-bin)/pkg/cmd/release/release_/release post-blockers \
  --release-series 22.1 \
  --smtp-user $MY_SMTP_USER --smtp-host smtp.gmail.com --smtp-port 587 \
  --to $MY_TO_EMAIL \
  --template-dir ./pkg/cmd/release/templates \
  --prep-date 2022-04-01 --publish-date 2022-05-01 \
  --next-version v22.1.0-alpha.3
```

```
# Test 2: generates email for stable release, no need to set `next-version` parameter
# --release-series 21.1
SMTP_PASSWORD=$MY_SMTP_PWD SMTP_USER=$MY_SMTP_USER GITHUB_TOKEN=$MY_GITHUB_TOKEN \
$(bazel info bazel-bin)/pkg/cmd/release/release_/release post-blockers \
  --release-series 21.1 \
  --smtp-user $MY_SMTP_USER --smtp-host smtp.gmail.com --smtp-port 587 \
  --to $MY_TO_EMAIL \
  --template-dir ./pkg/cmd/release/templates \
  --prep-date 2022-04-01 --publish-date 2022-05-01
```

Release note: None
Release justification: non-production code change.

### CODE REVIEW

**Comment by celiala** [comment]
Hi @jlinder @rail - this PR still needs to be tested (and TODOs implemented); but this is ready for a preliminary review, whenever you get a moment, thanks!

**Comment by celiala** [comment]
Hi @jlinder @rail - now that the email template has merged, this PR has been rebased / cleaned up, and is RFAL, thanks!

File: pkg/cmd/release/github.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	for {
+		events, _, err := c.client.Issues.ListIssueTimeline(
+			c.ctx, c.owner, c.repo, issueNum, &github.ListOptions{
+				PerPage: githubPageSize,
+				Page:    pageNum,
+			})
+		if err != nil {
+			return "", err
+		}
+		for i := len(events) - 1; i >= 0; i-- {
```
Comment (celiala) [inline pkg/cmd/release/github.go:137]: TODO - double-check this API:

I think this needs to start at i = 0 and increment by +1. Looking at this example, event[0] is the most recent, which is where I want to start:

https://api.github.com/repos/cockroachdb/cockroach/issues/66606/timeline

**Review by rail — CHANGES_REQUESTED** [review]
I have no hard blockers here, just some nits and questions/ideas. Marking as blocking to make sure we regenerate all the vendor stuff to make CI happy.
:dancers: 

> Reviewed 1 of 1 files at r2, 5 of 8 files at r3, 41 of 43 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-:-MxoRoQ88_JHD8fXtpkb:b-3mvqrt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @celiala and @jlinder)

___
*[DEPS.bzl, line 3463 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoFDqLDMnwOcbc5SvF:-MxoFDqLDMnwOcbc5SvG:b-69foi2) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/DEPS.bzl#L3463)):*
> ```Starlark
>         build_file_proto_mode = "disable_global",
>         importpath = "github.com/google/go-querystring",
>         # TODO: mirror this repo (to fix, run `./dev generate bazel --mirror`)
> ```

A gentle reminder to mirror the deps and push the updated code before merging.
___
*[go.mod, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoFRZj7KGkMAV_URaL:-MxoFRZj7KGkMAV_URaM:bc2ubiu) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/go.mod#L83)):*
> ```XML
> 	github.com/google/go-cmp v0.5.6
> 	github.com/google/go-github v17.0.0+incompatible
> 	github.com/google/go-github/v42 v42.0.0
> ```

CI complained about this added but not vendored. Need to do some vendor magic.
___
*[pkg/cmd/release/blockers.go, line 120 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoIPcn1MmDjumGxpjB:-MxoIPcn1MmDjumGxpjC:botobqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/blockers.go#L120)):*
> ```Go
> 		blockersURL = "go.crdb.dev/blockers"
> 		releaseBranch = "master"
> 		releaseBranchLabel = "branch-master"
> ```

Not a blocker. The block above implies that it will use `master` for `22.1` and `123.1`. The latter is a typo. I wonder how we can either fail or at least warn that we are using a fallback here? :thinking: 
___
*[pkg/cmd/release/github.go, line 77 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoLUMd5hddvRuSy3GQ:-MxoLUMd5hddvRuSy3GR:b-fmvhy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L77)):*
> ```Go
> 				ListOptions: github.ListOptions{
> 					PerPage: githubPageSize,
> 					Page:    pageNum,
> ```

Heh. No wonder why they switched to cursor based pagination in graphql. This is a bomb for a race condition :tongue: 
___
*[pkg/cmd/release/github.go, line 97 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoMIFTDflf_szRu4XW:-MxoMIFTDflf_szRu4XX:b69oohw) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L97)):*
> ```Go
> 		}
> 
> 		if len(issues) < githubPageSize {
> ```

I wonder if they have exact number of issues the search returns somewhere in the response. :thinking: GraphQL has `issueCount` or something like this to indicate the amount even in the first page.
___
*[pkg/cmd/release/github.go, line 137 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxDL9gh_DNAXCq9BWer-r5-137:-MxoOmbD3T05QZDcN5_U:b-2m78n) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L137)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

TODO - double-check this API:

I think this needs to start at i = 0 and increment by +1. Looking at this example, event[0] is the most recent, which is where I want to start:

https://api.github.com/repos/cockroachdb/cockroach/issues/66606/timeline


</blockquote></details>

I think the current approach should work assuming github appends the events to the timeline. I wonder if we can sort by `created_at` just to be sure...
___
*[pkg/cmd/release/github.go, line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoPAaY01VdD7Jauluh:-MxoPAaY01VdD7Jaului:b-9m2t21) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L143)):*
> ```Go
> 					project, _, err := c.client.Projects.GetProject(c.ctx, *projectId)
> 					if err != nil {
> 						return "", err
> ```

May I ask you to wrap the errors, so it's easier to find actual issues. Something like
```go
return "", fmt.Errorf("cannot fetch projects: %w", err)
```
___
*[pkg/cmd/release/github.go, line 147 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoQIzBAA0nq_BROe4I:-MxoQIzBAA0nq_BROe4J:bc77vi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L147)):*
> ```Go
> 					// removed event means that the ProjectCard was removed from the issue.
> 					if strings.Contains(*events[i].Event, "removed") {
> 						removedProjects[*project.Name] = true
> ```

Can you just `return "", nil` on this condition? To me it sounds like this is the most recent event and the project card is removed. :thinking: If it makes sense, you can get rid of `removedProjects`
___
*[pkg/cmd/release/jira.go, line 89 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoR2_O0blEmBl7pJuO:-MxoR2_O0blEmBl7pJuP:b9le3pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/jira.go#L89)):*
> ```Go
> 
> type jiraIssue struct {
>   ID           string
> ```

No idea why the spacing has changed here. :man_shrugging: Welp...

**Review by celiala — COMMENTED** [review]
TFTR! This is RFAL:

I still need to get the vendor / CI stuff worked out (last remaining thing to address from first round of feedback), but I wanted to push the remaining changes in now to get feedback on those.

Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-:-Mxu5JhT7rThQ57MK99J:b2katjm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[DEPS.bzl, line 3463 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoFDqLDMnwOcbc5SvF:-Mxu-rXy72x8bN8d2FjS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/DEPS.bzl#L3463)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

A gentle reminder to mirror the deps and push the updated code before merging.
</blockquote></details>

Done.
___
*[go.mod, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoFRZj7KGkMAV_URaL:-Mxu-vdJ7Q5H45cESamy:b-relgx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/go.mod#L83)):*
> CI complained about this added but not vendored. Need to do some vendor magic.

(working on this now)
___
*[pkg/cmd/release/blockers.go, line 120 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoIPcn1MmDjumGxpjB:-Mxu05qj75rEfFMgmoqU:bu1y1i4) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/blockers.go#L120)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Not a blocker. The block above implies that it will use `master` for `22.1` and `123.1`. The latter is a typo. I wonder how we can either fail or at least warn that we are using a fallback here? :thinking: 
</blockquote></details>

added simple check (see `if !strings.HasPrefix(blockersFlags.nextVersion..`)
___
*[pkg/cmd/release/blockers.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-Mxu02404yVuLWP_EgZB:-Mxu02404yVuLWP_EgZC:b-l8p4vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/54afd9e25095623672c942130b133c2111130361/pkg/cmd/release/blockers.go#L112)):*
> Not a blocker. The block above implies that it will use master for 22.1 and 123.1. The latter is a typo. I wonder how we can either fail or at least warn that we are using a fallback here? 🤔

added simple check here for this
___
*[pkg/cmd/release/github.go, line 77 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoLUMd5hddvRuSy3GQ:-Mxu37qb6xIasdWPXcox:bazgqru) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L77)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Heh. No wonder why they switched to cursor based pagination in graphql. This is a bomb for a race condition :tongue: 
</blockquote></details>

Added TODO/notes (for future)
___
*[pkg/cmd/release/github.go, line 97 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoMIFTDflf_szRu4XW:-Mxu3FRIBlDS8GsFtj10:boed9qd) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L97)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

I wonder if they have exact number of issues the search returns somewhere in the response. :thinking: GraphQL has `issueCount` or something like this to indicate the amount even in the first page.
</blockquote></details>

That would be nice -- unfortunately, neither this go package nor the raw API return a `totalCount` :/
___
*[pkg/cmd/release/github.go, line 137 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxDL9gh_DNAXCq9BWer-r5-137:-Mxu3_q67IDYAiI_o1Yy:b-15ugsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L137)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

I think the current approach should work assuming github appends the events to the timeline. I wonder if we can sort by `created_at` just to be sure...
</blockquote></details>

Ack: sort added

Note that I refactored this method to separate out the API logic (kept in `github.go`) and the business logic (put into new `mostRecentProjectName()`method in `blockers.go`). 

The sort is in the `mostRecentProjectName()`.
___
*[pkg/cmd/release/github.go, line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoPAaY01VdD7Jauluh:-Mxu494jCj45w5IQ9jtD:beb5sbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L143)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

May I ask you to wrap the errors, so it's easier to find actual issues. Something like
```go
return "", fmt.Errorf("cannot fetch projects: %w", err)
```
</blockquote></details>

doh, my bad. done! :)
___
*[pkg/cmd/release/github.go, line 147 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoQIzBAA0nq_BROe4I:-Mxu4FM2AkpNUWrkcTPI:b8wrfwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/github.go#L147)):*
> Can you just return "", nil on this condition? 

Unfortunately, no ...

this method goes in reverse chronological order, since we want to get the most recently-added project. However, if we see a "remove project X" event on April 15, that means we should ignore the "add project X" event on April 14, because it was subsequently removed.

To make this a bit cleaner, I pulled the API logic and business logic into 2 separate functions. Hopefully, the above makes more sense by looking at `mostRecentProjectName()` and the corresponding unit test.
___
*[pkg/cmd/release/github.go, line 79 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-Mxu2wZH9AlcHMSeA80s:-Mxu2wZH9AlcHMSeA80t:b-uvfv3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bbbcfa68b50d213bb7ef7fbc6c9a0e262c215e6/pkg/cmd/release/github.go#L79)):*
> Heh. No wonder why they switched to cursor based pagination in graphql. This is a bomb for a race condition 👅

added note here, for potentially switching to graphql api in the future.
___
*[pkg/cmd/release/github.go, line 142 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-Mxu365b3YiOW_yCrpUw:-Mxu365b3YiOW_yCrpUx:b-uvfv3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bbbcfa68b50d213bb7ef7fbc6c9a0e262c215e6/pkg/cmd/release/github.go#L142)):*
> Heh. No wonder why they switched to cursor based pagination in graphql. This is a bomb for a race condition 👅

added note here, for potentially switching to graphql api in the future.
___
*[pkg/cmd/release/jira.go, line 89 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-MxoR2_O0blEmBl7pJuO:-Mxu4xtG-MRrxJrdyVe1:b-whtzkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/202a67f1ed561cf7ca0958d11a7ab9a85feab0e6/pkg/cmd/release/jira.go#L89)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

No idea why the spacing has changed here. :man_shrugging: Welp...
</blockquote></details>

I think I don't have my GoLand preferences quite set up right (I used to have some auto-gofmt & maybe tab/spacing configs in place)?

**Review by rail — APPROVED** [review]
:lgtm:
Just need to reflect the `vendor` stuff.

> Reviewed 4 of 5 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77250#-:-MxuDFFIE6mH-Nl3JlFS:bbts41m)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala and @jlinder)

**Comment by celiala** [comment]
TFTR!

bors r=rail

**Pull request merged**